File: spriteforest-ios/action/SFActionEase.m
//
//  SFActionEase.m
//  SpriteForest2
//
//  Created by Moky on 12-10-2.
//  Copyright 2012 Slanissue.com. All rights reserved.
//

#import "SFActionEase.h"

#import "cnMacros.h"


@implementation SFActionEase

@synthesize action = action_;

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		id action = [[self class] subActionFromDictionary:dict];
		if ([action isKindOfClass:[CCActionInterval class]])
		{
			self.action = action;
		}
	}
	return self;
}

- (id) init
{
	if ((self = [super init]))
	{
		self.action = nil;
	}
	return self;
}

- (void) dealloc
{
	[action_ release];
	[super dealloc];
}

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCActionEase actionWithAction:action_];
}

@end


#pragma mark - Action Ease Rate


@implementation SFEaseRateAction

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		rate_ = [[dict objectForKey:@"Rate"] floatValue];
	}
	return self;
}

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseRateAction actionWithAction:action_ rate:rate_];
}

@end


@implementation SFEaseIn

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseIn actionWithAction:action_ rate:rate_];
}

@end


@implementation SFEaseOut

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseOut actionWithAction:action_ rate:rate_];
}

@end


@implementation SFEaseInOut

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseInOut actionWithAction:action_ rate:rate_];
}

@end


#pragma mark - Action Ease Exponential


@implementation SFEaseExponentialIn

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseExponentialIn actionWithAction:action_];
}

@end


@implementation SFEaseExponentialOut

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseExponentialOut actionWithAction:action_];
}

@end


@implementation SFEaseExponentialInOut

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseExponentialInOut actionWithAction:action_];
}

@end


#pragma mark - Action Ease Sine


@implementation SFEaseSineIn

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseSineIn actionWithAction:action_];
}

@end


@implementation SFEaseSineOut

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseSineOut actionWithAction:action_];
}

@end


@implementation SFEaseSineInOut

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseSineInOut actionWithAction:action_];
}

@end


#pragma mark - Action Ease Elastic


@implementation SFEaseElastic

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseElastic actionWithAction:action_ period:duration_];
}

@end


@implementation SFEaseElasticIn

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseElasticIn actionWithAction:action_ period:duration_];
}

@end


@implementation SFEaseElasticOut

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseElasticOut actionWithAction:action_ period:duration_];
}

@end


@implementation SFEaseElasticInOut

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseElasticInOut actionWithAction:action_ period:duration_];
}

@end


#pragma mark - Action Ease Bounce


@implementation SFEaseBounce

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseBounce actionWithAction:action_];
}

@end


@implementation SFEaseBounceIn

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseBounceIn actionWithAction:action_];
}

@end


@implementation SFEaseBounceOut

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseBounceOut actionWithAction:action_];
}

@end


@implementation SFEaseBounceInOut

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseBounceInOut actionWithAction:action_];
}

@end


#pragma mark - Action Ease Back


@implementation SFEaseBackIn

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseBackIn actionWithAction:action_];
}

@end


@implementation SFEaseBackOut

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseBackOut actionWithAction:action_];
}

@end


@implementation SFEaseBackInOut

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: action = %@", action_);
		return [super ccAction];
	}
	return [CCEaseBackInOut actionWithAction:action_];
}

@end

File: x/action/SFActionEase.cpp
//
//  SFActionEase.cpp
//  SpriteForest-X
//
//  Created by Moky on 12-11-29.
//  Copyright 2012 Slanissue Inc. All rights reserved.
//

#include "SFActionEase.h"

NS_SF_BEGIN

USING_NS_CC;
USING_NS_CN;

#define SF_IMPLEMENT_CCACTION_WITH_ACTION(action_name) /* ccAction: actionWithAction(pAction) begin */ \
CCAction * SF##action_name::ccAction(void) { \
	CCActionInterval * pAction = getAction(); \
	return pAction ? CC##action_name::actionWithAction(pAction) : NULL; \
} /* ccAction - actionWithAction(pAction) end */

#define SF_IMPLEMENT_CCACTION_WITH_ACTION_RATE(action_name) /* ccAction: actionWithAction(pAction, fRate) begin */ \
CCAction * SF##action_name::ccAction(void) { \
	CCActionInterval * pAction = getAction(); \
	return pAction ? CC##action_name::actionWithAction(pAction, m_fRate) : NULL; \
} /* ccAction - actionWithAction(pAction, fRate) end */

#define SF_IMPLEMENT_CCACTION_WITH_ACTION_DURATION(action_name) /* ccAction: actionWithAction(pAction, fRate) begin */ \
CCAction * SF##action_name::ccAction(void) { \
	CCActionInterval * pAction = getAction(); \
	return pAction ? CC##action_name::actionWithAction(pAction, m_fDuration) : NULL; \
} /* ccAction - actionWithAction(pAction, fRate) end */



// SFActionEase
SFActionEase::SFActionEase(void)
: SFActionInterval()
, m_pAction(NULL)
{
	
}

SFActionEase::~SFActionEase(void)
{
	setAction(NULL);
}

bool SFActionEase::init(void)
{
	setAction(NULL);
	return true;
}

bool SFActionEase::initWithDictionary(CCDictionary * pDict)
{
	if (!SFActionInterval::initWithDictionary(pDict))
	{
		CNLog("error");
		return false;
	}
	
	CCAction * pAction = SFActionEase::subActionFromDictionary(pDict);
	setAction(dynamic_cast<CCActionInterval *>(pAction));
	
	return true;
}

SF_IMPLEMENT_CCACTION_WITH_ACTION(ActionEase);

#pragma mark - Action Ease Rate

// SFEaseRateAction
bool SFEaseRateAction::initWithDictionary(CCDictionary * pDict)
{
	if (!SFActionEase::initWithDictionary(pDict))
	{
		CNLog("error");
		return false;
	}
	
	CCString * pRate = (CCString *)pDict->objectForKey("Rate");
	m_fRate = pRate ? pRate->floatValue() : 0;
	
	return true;
}

SF_IMPLEMENT_CCACTION_WITH_ACTION_RATE(EaseRateAction);

// SFEaseIn
SF_IMPLEMENT_CCACTION_WITH_ACTION_RATE(EaseIn);

// SFEaseOut
SF_IMPLEMENT_CCACTION_WITH_ACTION_RATE(EaseOut);

// SFEaseInOut
SF_IMPLEMENT_CCACTION_WITH_ACTION_RATE(EaseInOut);

#pragma mark - Action Ease Exponential

// SFEaseExponentialIn
SF_IMPLEMENT_CCACTION_WITH_ACTION(EaseExponentialIn);

// SFEaseExponentialOut
SF_IMPLEMENT_CCACTION_WITH_ACTION(EaseExponentialOut);

// SFEaseExponentialInOut
SF_IMPLEMENT_CCACTION_WITH_ACTION(EaseExponentialInOut);

#pragma mark - Action Ease Sine

// SFEaseSineIn
SF_IMPLEMENT_CCACTION_WITH_ACTION(EaseSineIn);

// SFEaseSineOut
SF_IMPLEMENT_CCACTION_WITH_ACTION(EaseSineOut);

// SFEaseSineInOut
SF_IMPLEMENT_CCACTION_WITH_ACTION(EaseSineInOut);

#pragma mark - Action Ease Elastic

// SFEaseElastic
SF_IMPLEMENT_CCACTION_WITH_ACTION_DURATION(EaseElastic);

// SFEaseElasticIn
SF_IMPLEMENT_CCACTION_WITH_ACTION_DURATION(EaseElasticIn);

// SFEaseElasticOut
SF_IMPLEMENT_CCACTION_WITH_ACTION_DURATION(EaseElasticOut);

// SFEaseElasticInOut
SF_IMPLEMENT_CCACTION_WITH_ACTION_DURATION(EaseElasticInOut);

#pragma mark - Action Ease Bounce

// SFEaseBounce
SF_IMPLEMENT_CCACTION_WITH_ACTION(EaseBounce);

// SFEaseBounceIn
SF_IMPLEMENT_CCACTION_WITH_ACTION(EaseBounceIn);

// SFEaseBounceOut
SF_IMPLEMENT_CCACTION_WITH_ACTION(EaseBounceOut);

// SFEaseBounceInOut
SF_IMPLEMENT_CCACTION_WITH_ACTION(EaseBounceInOut);

#pragma mark - Action Ease Back

// SFEaseBackIn
SF_IMPLEMENT_CCACTION_WITH_ACTION(EaseBackIn);

// SFEaseBackOut
SF_IMPLEMENT_CCACTION_WITH_ACTION(EaseBackOut);

// SFEaseBackInOut
SF_IMPLEMENT_CCACTION_WITH_ACTION(EaseBackInOut);


NS_SF_END
